README.rst
********************
collective.js.cycle2
********************

.. contents:: Table of Contents

Life, the Universe, and Everything
==================================

`Cycle2 <http://jquery.malsup.com/cycle2/>`_ is a versatile slideshow plugin for jQuery built around ease-of-use.
It supports a declarative initialization style that allows full customization without any scripting.
Simply include the plugin, declare your markup, and Cycle2 does the rest.

This package adds a browser resource to use Cycle2 and its plugins on a Plone site.

Mostly Harmless
===============

.. image:: http://img.shields.io/pypi/v/collective.js.cycle2.svg
    :target: https://pypi.python.org/pypi/collective.js.cycle2

.. image:: https://img.shields.io/travis/collective/collective.js.cycle2/master.svg
    :target: http://travis-ci.org/collective/collective.js.cycle2

.. image:: https://img.shields.io/coveralls/collective/collective.js.cycle2/master.svg
    :target: https://coveralls.io/r/collective/collective.js.cycle2

Got an idea? Found a bug? Let us know by `opening a support ticket <https://github.com/collective/collective.js.cycle2/issues>`_.

Don't Panic
===========

Installation
------------

To enable this package in a buildout-based installation:

#. Edit your buildout.cfg and add add the following to it:

.. code-block:: ini

    [buildout]
    ...
    eggs =
        collective.js.cycle2

After updating the configuration you need to run ''bin/buildout'', which will take care of updating your system.

Go to the 'Site Setup' page in a Plone site and click on the 'Add-ons' link.

Check the box next to ``collective.js.cycle2`` and click the 'Activate' button.

Usage
-----

If your page template inherits from ``main_template``,
just include the resources on it by usign the following syntax:

.. code-block:: xml

    <metal:block fill-slot="javascript_head_slot">
      <script tal:attributes="src string:$portal_url/++resource++collective.js.cycle2/jquery.cycle2.min.js"></script>
    </metal:block>

Alternatively you can add them into your site's JavaScript Registry directly or by using GenericSetup:

.. code-block:: xml

    <?xml version="1.0"?>
    <object name="portal_javascripts">
      <javascript
          cacheable="True" compression="none" cookable="True" enabled="True"
          id="++resource++collective.js.cycle2/jquery.cycle2.min.js" />
    </object>

Plugins
-------

The package also includes the code for the following plugins:

* Transition
    * Carousel
    * Flip
    * IE-Fade
    * ScrollVert
    * Shuffle
    * Tile
* Functional
    * Caption2
    * Center
    * Swipe
    * YouTube

Check Cycle2 `download page <http://jquery.malsup.com/cycle2/download/>`_ for more information.

Utility
=======

The ``utils.js`` script has an utility that:

* fix the player size according with the width of the container and the aspect ratio
* vertically center images
* sync the player, description and thumbnails when click in other picture or next/prev buttons

Usage
-----

To use the utility, you need to add the script in the same way you add Cycle2 resources:

If your page template inherits from ``main_template``,
just include the script on it by usign the following syntax:

.. code-block:: xml

    <metal:block fill-slot="javascript_head_slot">
      <script tal:attributes="src string:$portal_url/++resource++collective.js.cycle2/utils.min.js"></script>
    </metal:block>

Alternatively you can add it directly into your site's JavaScript Registry or by using GenericSetup:

.. code-block:: xml

    <?xml version="1.0"?>
    <object name="portal_javascripts">
      <javascript
          cacheable="True" compression="none" cookable="True" enabled="True"
          id="++resource++collective.js.cycle2/utils.min.js" />
    </object>

And in your script you should call the utility object passing the gallery element:

.. code-block:: javascript

    $(window).load(function() {
      var i, len, ref, slideshow;
      ref = $('.slideshow-container');
      for (i = 0, len = ref.length; i < len; i++) {
        slideshow = ref[i];
        new cycle2SlideShow(slideshow);
      }
    });

The script is currently used in `sc.photogallery <https://github.com/simplesconsultoria/sc.photogallery>`_ and  `collective.nitf <https://github.com/collective/collective.nitf>`_.

Not entirely unlike
===================

`collective.js.galleria <https://pypi.python.org/pypi/collective.js.galleria>`_
    Galleria is a JavaScript image gallery framework built on top of the jQuery library.
    The aim is to simplify the process of creating professional image galleries for the web and mobile devices.


CHANGES.rst
Changelog
=========

There's a frood who really knows where his towel is.

1.0rc1 (unreleased)
-------------------

- Nothing changed yet.


1.0b2 (2016-07-20)
------------------

- Fade in slideshows to avoid showing them before Cycle2 is loaded.
  [rodfersou]

- Add `utils.js` script;
  this script can be used by third party add-ons to easily create responsive slideshows.
  [rodfersou]


1.0b1 (2015-05-27)
------------------

- Add control panel configlet to define global options for the Cycle2 slideshow plugin (closes `#3`_).
  [hvelarde]

- Remove browser layer and JS resource registry registrations.
  Include all Cycle2 and plugins code as browser resources (closes `#1`_).
  [rodfersou]


1.0a1 (2014-10-20)
------------------

- Initial release.

.. _`#1`: https://github.com/collective/collective.js.cycle2/issues/1
.. _`#3`: https://github.com/collective/collective.js.cycle2/issues/3


src/collective/js/cycle2/tests/test_setup.py
# -*- coding: utf-8 -*-
from collective.js.cycle2.config import PROJECTNAME
from collective.js.cycle2.testing import INTEGRATION_TESTING

import unittest


class InstallTestCase(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        self.qi = self.portal['portal_quickinstaller']

    def test_installed(self):
        self.assertTrue(self.qi.isProductInstalled(PROJECTNAME))


class UninstallTestCase(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        self.qi = self.portal['portal_quickinstaller']
        self.qi.uninstallProducts(products=[PROJECTNAME])

    def test_uninstalled(self):
        self.assertFalse(self.qi.isProductInstalled(PROJECTNAME))


src/collective/js/cycle2/tests/test_upgrades.py
# -*- coding: utf-8 -*-
from collective.js.cycle2.interfaces import IAddOnInstalled
from collective.js.cycle2.testing import INTEGRATION_TESTING
from plone import api
from plone.browserlayer import utils
from zope.interface import Interface

import unittest


class UpgradeTestCaseBase(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self, from_version, to_version):
        self.portal = self.layer['portal']
        self.setup = self.portal['portal_setup']
        self.profile_id = u'collective.js.cycle2:default'
        self.from_version = from_version
        self.to_version = to_version

    def _get_upgrade_step(self, title):
        """Get one of the upgrade steps.

        Keyword arguments:
        title -- the title used to register the upgrade step
        """
        self.setup.setLastVersionForProfile(self.profile_id, self.from_version)
        upgrades = self.setup.listUpgrades(self.profile_id)
        steps = [s for s in upgrades[0] if s['title'] == title]
        return steps[0] if steps else None

    def _do_upgrade_step(self, step):
        """Execute an upgrade step.

        Keyword arguments:
        step -- the step we want to run
        """
        request = self.layer['request']
        request.form['profile_id'] = self.profile_id
        request.form['upgrades'] = [step['id']]
        self.setup.manage_doUpgrades(request=request)

    def _how_many_upgrades_to_do(self):
        self.setup.setLastVersionForProfile(self.profile_id, self.from_version)
        upgrades = self.setup.listUpgrades(self.profile_id)
        assert len(upgrades) > 0
        return len(upgrades[0])


class Upgrade1000to1001TestCase(UpgradeTestCaseBase):

    def setUp(self):
        UpgradeTestCaseBase.setUp(self, u'1000', u'1001')

    def test_upgrade_to_1001_registrations(self):
        version = self.setup.getLastVersionForProfile(self.profile_id)[0]
        self.assertGreaterEqual(int(version), int(self.to_version))
        self.assertEqual(self._how_many_upgrades_to_do(), 1)

    def test_issue_1(self):
        # check if the upgrade step is registered
        title = u'Remove browser layer and jsregistry'
        description = u'We don\'t need a profile for this package.'
        step = self._get_upgrade_step(title)
        self.assertIsNotNone(step)
        self.assertEqual(step['description'], description)

        js_tool = self.portal['portal_javascripts']
        JS_IDS = [
            '++resource++collective.js.cycle2/jquery.cycle2.min.js',
            '++resource++collective.js.cycle2/jquery.cycle2.center.min.js',
            '++resource++collective.js.cycle2/jquery.cycle2.swipe.min.js'
        ]
        for js_id in JS_IDS:
            js_tool.registerResource(js_id)

        resource_ids = js_tool.getResourceIds()
        for js_id in JS_IDS:
            self.assertIn(js_id, resource_ids)

        utils.register_layer(IAddOnInstalled, name='collective.js.cycle2')
        self.assertIn(IAddOnInstalled, utils.registered_layers())

        self._do_upgrade_step(step)

        resource_ids = js_tool.getResourceIds()
        for js_id in JS_IDS:
            self.assertNotIn(js_id, resource_ids)

        self.assertNotIn(IAddOnInstalled, utils.registered_layers())


src/collective/js/cycle2/tests/test_controlpanel.py
# -*- coding: utf-8 -*-
from collective.js.cycle2.config import PROJECTNAME
from collective.js.cycle2.controlpanel import ICycle2Settings
from collective.js.cycle2.testing import INTEGRATION_TESTING
from plone import api
from plone.app.testing import logout
from plone.registry.interfaces import IRegistry
from zope.component import getUtility

import unittest


class ControlPanelTestCase(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        self.controlpanel = self.portal['portal_controlpanel']

    def test_controlpanel_has_view(self):
        request = self.layer['request']
        view = api.content.get_view(u'cycle2-settings', self.portal, request)
        view = view.__of__(self.portal)
        self.assertTrue(view())

    def test_controlpanel_view_is_protected(self):
        from AccessControl import Unauthorized
        logout()
        with self.assertRaises(Unauthorized):
            self.portal.restrictedTraverse('@@cycle2-settings')

    def test_controlpanel_installed(self):
        actions = [a.getAction(self)['id']
                   for a in self.controlpanel.listActions()]
        self.assertIn('cycle2', actions, 'control panel not installed')

    def test_controlpanel_removed_on_uninstall(self):
        qi = self.portal['portal_quickinstaller']

        with api.env.adopt_roles(['Manager']):
            qi.uninstallProducts(products=[PROJECTNAME])

        actions = [a.getAction(self)['id']
                   for a in self.controlpanel.listActions()]
        self.assertNotIn('cycle2', actions, 'control panel not removed')


class RegistryTestCase(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        self.registry = getUtility(IRegistry)
        self.settings = self.registry.forInterface(ICycle2Settings)

    def test_paused_record_in_registry(self):
        self.assertTrue(hasattr(self.settings, 'paused'))
        self.assertEqual(self.settings.paused, True)

    def test_allow_wrap_record_in_registry(self):
        self.assertTrue(hasattr(self.settings, 'allow_wrap'))
        self.assertEqual(self.settings.allow_wrap, True)

    def test_caption_template_record_in_registry(self):
        self.assertTrue(hasattr(self.settings, 'caption_template'))
        self.assertEqual(self.settings.caption_template, u'{{slideNum}} / {{slideCount}}')

    def test_pager_template_record_in_registry(self):
        self.assertTrue(hasattr(self.settings, 'pager_template'))
        self.assertEqual(self.settings.pager_template, u'<span>&bull;</span>')

    def test_speed_record_in_registry(self):
        self.assertTrue(hasattr(self.settings, 'speed'))
        self.assertEqual(self.settings.speed, 500)

    def test_timeout_record_in_registry(self):
        self.assertTrue(hasattr(self.settings, 'timeout'))
        self.assertEqual(self.settings.timeout, 4000)

    def test_transition_effect_record_in_registry(self):
        self.assertTrue(hasattr(self.settings, 'transition_effect'))
        self.assertEqual(self.settings.transition_effect, 'fade')

    def test_default_image_size_record_in_registry(self):
        self.assertTrue(hasattr(self.settings, 'default_image_size'))
        self.assertEqual(self.settings.default_image_size, 'preview')

    def test_records_removed_on_uninstall(self):
        qi = self.portal['portal_quickinstaller']

        with api.env.adopt_roles(['Manager']):
            qi.uninstallProducts(products=[PROJECTNAME])

        records = [
            ICycle2Settings.__identifier__ + '.paused',
            ICycle2Settings.__identifier__ + '.allow_wrap',
            ICycle2Settings.__identifier__ + '.caption_template',
            ICycle2Settings.__identifier__ + '.pager_template',
            ICycle2Settings.__identifier__ + '.speed',
            ICycle2Settings.__identifier__ + '.timeout',
            ICycle2Settings.__identifier__ + '.transition_effect',
            ICycle2Settings.__identifier__ + '.default_image_size',
        ]

        for r in records:
            self.assertNotIn(r, self.registry)


src/collective/js/cycle2/resources/utils.js
(function(window, factory) {
  var cycle2SlideShow = factory(window, window.document);
  window.cycle2SlideShow = cycle2SlideShow;
  if (typeof module == 'object' && module.exports){
    module.exports = cycle2SlideShow;
  }
}
(window, function l(window, document) {
  'use strict';
  var Cycle2SlideShow = (function() {
    function Cycle2SlideShow(el) {
      $(el).data('cycle2slideshow', this);
      this.$player = $('.cycle-player', el);
      this.$thumbsContainer = $('.cycle-carrossel', el);
      this.$thumbs = $('.cycle-carrossel .thumb-itens', el)
      this.$imgs = $('.cycle-player img', el);
      this.$slideshows = $('.cycle-slideshow', el);
      this.aspectRatio = parseFloat($(el).attr('data-aspect-ratio'));
      if (isNaN(this.aspectRatio)) {
        this.aspectRatio = 3 / 2;
      }
      this.maxWidth = this.$player.width();
      this.maxHeight = this.maxWidth / this.aspectRatio;
      this.bindEvents();
      this.fixImageSize();
      this.centerImage();
      $(el).fadeTo('slow', 1);
    }
    Cycle2SlideShow.prototype.bindEvents = function() {
      this.$player.on('cycle-next cycle-prev', $.proxy(this.syncSlideshows, this));
      this.$player.on('cycle-before', $.proxy(this.cycleBefore, this));
      this.$thumbs.on('click', $.proxy(this.thumbsClick, this));
    };
    Cycle2SlideShow.prototype.fixImageSize = function() {
      this.$player.css({
        'width': this.maxWidth,
        'height': this.maxHeight
      });
      this.$imgs.css({
        'max-width': this.maxWidth,
        'max-height': this.maxHeight
      });
    };
    Cycle2SlideShow.prototype.centerImage = function(img) {
      if (typeof(img) === 'undefined') {
        var img = this.$imgs[0];
        if ($(img).parent().hasClass('cycle-sentinel')) {
          img = this.$imgs[1];
        }
      }
      var img_height = parseInt($(img).attr('height'));
      if (img_height < this.maxHeight) {
        var margin = parseInt((this.maxHeight - img_height) / 2);
        $(img).css('margin-top', margin);
      }
    };
    Cycle2SlideShow.prototype.syncSlideshows = function(e, opts) {
      e.preventDefault();
      this.$slideshows.cycle('goto', opts.currSlide);
    };
    Cycle2SlideShow.prototype.cycleBefore = function(e, opts, outSlide, inSlide) {
      e.preventDefault();
      this.centerImage(inSlide.children[0]);
    };
    Cycle2SlideShow.prototype.thumbsClick = function(e) {
      e.preventDefault();
      var index = this.$thumbsContainer.data('cycle.API').getSlideIndex(e.currentTarget);
      this.$slideshows.cycle('goto', index);
    };
    return Cycle2SlideShow;
  })();
  return Cycle2SlideShow;
}));


src/collective/js/cycle2/controlpanel.py
# -*- coding: utf-8 -*-
from collective.js.cycle2 import _
from plone.app.registry.browser import controlpanel
from plone.directives import form
from zope import schema
from zope.schema.vocabulary import SimpleTerm
from zope.schema.vocabulary import SimpleVocabulary

fx = ['fade', 'fadeout', 'none', 'scrollHorz']
fx = SimpleVocabulary([SimpleTerm(f) for f in fx])

# TODO: get this from portal_properties/imaging_properties
scales = ['large', 'preview', 'mini', 'thumb', 'tile', 'icon', 'listing']
scales = SimpleVocabulary([SimpleTerm(s) for s in scales])


class ICycle2Settings(form.Schema):

    """Interface for the control panel form."""

    paused = schema.Bool(
        title=_(u'Paused'),
        description=_(u'If selected the slideshow will begin in a paused state.'),
        default=True,
    )

    allow_wrap = schema.Bool(
        title=_(u'Allow wrap'),
        description=_(u'Determines whether or not a slideshow can advance from the last slide to the first (or vice versa).'),
        default=True,
    )

    caption_template = schema.TextLine(
        title=_(u'Caption Template'),
        description=_(
            u'A template string which defines how the slideshow caption should be formatted. '
            u'',
        ),
        default=u'{{slideNum}} / {{slideCount}}',
    )

    pager_template = schema.TextLine(
        title=_(u'Pager Template'),
        description=_(
            u'A template string which defines how the pager links should be formatted. '
            u'',
        ),
        default=u'<span>&bull;</span>',
    )

    speed = schema.Int(
        title=_(u'Speed'),
        description=_(u'The speed of the transition effect in milliseconds.'),
        default=500,
    )

    timeout = schema.Int(
        title=_(u'Timeout'),
        description=_(u'The time between slide transitions in milliseconds.'),
        default=4000,
    )

    transition_effect = schema.Choice(
        title=_(u'Transition effect'),
        description=_(u'The name of the slideshow transition to use.'),
        required=True,
        vocabulary=fx,
        default='fade',
    )

    default_image_size = schema.Choice(
        title=_(u'Default Image Size'),
        description=_(u''),
        required=True,
        vocabulary=scales,
        default='preview',
    )


class Cycle2SettingsEditForm(controlpanel.RegistryEditForm):
    schema = ICycle2Settings
    label = _(u'Cycle2')
    description = _(
        u'Here you can define global configuration options for the Cycle2 slideshow plugin.')


class Cycle2SettingsControlPanel(controlpanel.ControlPanelFormWrapper):
    form = Cycle2SettingsEditForm
